pkg/apis/core/validation/validation.go

```diff
@@ -4958,6 +4969,27 @@ func validateStorageNodeAffinityAnnotation(annotations map[string]string, fldPat
 	return policySpecified, allErrs
 }
 
+// validateVolumeNodeAffinity tests that the PersistentVolume.NodeAffinity has valid data
+func validateVolumeNodeAffinity(nodeAffinity *core.VolumeNodeAffinity, fldPath *field.Path) (bool, field.ErrorList) {
+	allErrs := field.ErrorList{}
+
+	if nodeAffinity == nil {
+		return false, allErrs
+	}
+
+	if !utilfeature.DefaultFeatureGate.Enabled(features.VolumeScheduling) {
+		allErrs = append(allErrs, field.Forbidden(fldPath, "Volume node affinity is disabled by feature-gate"))
+	}
+
+	if nodeAffinity.Required != nil {
+		allErrs = append(allErrs, ValidateNodeSelector(nodeAffinity.Required, fldPath.Child("required"))...)
```

Comment:
I'm in the middle of designing dynamic provisioning with topology and looking at how that's going to work with CSI.  The way that topology is specified in the CSI spec is not going to be able to support the full functionality of a NodeSelector.  The CSI spec represents topology as a map<string>[]string, where the key is the label key, and the value is an array of label values.  This is sufficient to support topologies of all the storage systems I can think of.

But I think from a Kubernetes standpoint, there is value in adopting consistent structs to describe topology across the system, even if it can't be fully supported.  So I still prefer representing volume topology with NodeAffinity in the Kubernetes API, and I could add validation here to only allow for the "In" operator.

@thockin @saad-ali @jsafrane and anyone else, would appreciate your thoughts

Reply:
Which semantics are not expressible?  Is CSI just "equals" or is it "in set".  What are the gaps?  "not in"?  What else?

What facet of CSI is this touching?  It's been a while since I looked at CSI in depth, but this starts to feel like it might be scope creep.

Reply:
Basically all the other node selector operators don't really make sense to describe volume accessibility: "NotIn", "Exists", "DoesNotExist", "Gt", "Lt".

For just the specification of volume topology + scheduling of those volumes (which is this PR), CSI can support this (the csi-kubernetes adapter can translate from csi topology -> k8s volume topology).  But once we consider dynamic provisioning, we need to do the translation the other way (k8s topology -> csi topology for the csi volume create call).  Agree that the CSI problem doesn't directly impact this PR, however, any new topology APIs we may have to add for dynamic provisioning could be inconsistent with how topology is specified here.

Reply:
I can make up use-cases for "not-in", but they can probably be solved by reciprocal "in".

Can you walk me through a hypothetical where we might use parts of NodeAffinity that CSI can't halndle?

Reply:
We discussed the use case of restricting topology in the StorageClass, moving it out from a plugin-specific parameter to a first class field, and then having to pass it to the provisioners.  In that use case, a NodeSelector doesn't make sense, and it's ok for the topology specification to be different as long as the underlying foundation of specifying topology keys and using node labels is similar.

So conclusion is: PV.NodeAffinity is fine the way it is.  Topology for dynamic provisioning may be more limited and can be specified using different APIs (and will be initially alpha).

Reply:
If no one else has concerns, then this PR is ready to go.